src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,178 @@
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        ///<inheritdoc cref="IBraceCompletionService.IsValidForBraceCompletionAsync(char, int, Document, CancellationToken)"/>
+        public virtual async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+        {
+            // check that the user is not typing in a string literal or comment
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
+        }
+
+        ///<inheritdoc cref="IBraceCompletionService.IsInsideCompletedBracesAsync(int, Document, CancellationToken)"/>
+        public async Task<BraceCompletionContext?> IsInsideCompletedBracesAsync(int caretLocation, Document document, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var leftToken = root.FindTokenOnLeftOfPosition(caretLocation);
+            var rightToken = root.FindTokenOnRightOfPosition(caretLocation);
+
+            if (IsValidOpeningBraceToken(leftToken) && IsValidClosingBraceToken(rightToken))
+            {
+                return new BraceCompletionContext(document, leftToken.GetLocation().SourceSpan.Start, rightToken.GetLocation().SourceSpan.End, caretLocation);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Checks if the already inserted token is a valid opening token at the position in the document.
+        /// </summary>
+        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            if (!IsParentSkippedTokensTrivia(syntaxFactsService, token))
+            {
+                return SpecializedTasks.False;
+            }
+
+            return Task.FromResult(IsValidOpeningBraceToken(token) && token.SpanStart == position);
+        }
+
+        /// <summary>
+        /// Checks that the current position is a valid location.
+        /// Used to determine if overtype should be allowed.
+        /// </summary>
+        protected static async Task<bool> CheckCurrentPositionAsync(Document document, int? currentPosition, CancellationToken cancellationToken)
+        {
+            // make sure auto closing is called from a valid position
+            if (currentPosition == null)
```

Comment:
this feels weird.  who is calling this with a possibly null position?  wouldn't they just bail out?  when is a null positoin ok to proceed?  when does this happen?